ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldStateDownloader.java

```diff
@@ -100,32 +98,35 @@ public WorldStateDownloader(
   }
 
   public CompletableFuture<Void> run(final BlockHeader header) {
-    LOG.info(
-        "Begin downloading world state from peers for block {} ({})",
-        header.getNumber(),
-        header.getHash());
     synchronized (this) {
-      if (status == Status.DONE || status == Status.RUNNING) {
-        return future;
+      if (isRunning()) {
+        CompletableFuture<Void> failed = new CompletableFuture<>();
+        failed.completeExceptionally(
+            new IllegalStateException(
+                "Cannot run an already running " + this.getClass().getSimpleName()));
+        return failed;
       }
-      status = Status.RUNNING;
-      future = new CompletableFuture<>();
-    }
+      LOG.info(
+          "Begin downloading world state from peers for block {} ({})",
+          header.getNumber(),
+          header.getHash());
+      future = createFuture();
 
-    Hash stateRoot = header.getStateRoot();
-    if (worldStateStorage.isWorldStateAvailable(stateRoot)) {
-      // If we're requesting data for an existing world state, we're already done
-      markDone();
-    } else {
-      pendingRequests.enqueue(NodeDataRequest.createAccountDataRequest(stateRoot));
-      requestNodeData(header);
+      Hash stateRoot = header.getStateRoot();
+      if (worldStateStorage.isWorldStateAvailable(stateRoot)) {
+        // If we're requesting data for an existing world state, we're already done
+        markDone();
+      } else {
+        pendingRequests.enqueue(NodeDataRequest.createAccountDataRequest(stateRoot));
+      }
     }
 
+    requestNodeData(header);
     return future;
   }
 
   public void cancel() {
-    // TODO
+    getFuture().cancel(true);
```

Comment:
Given that the use case for this is likely to be cancelling one download and immediately starting a new one for a later block, how do we avoid a race condition where outstanding requests and/or `handleCancellation` are still happening after this method returns and a new download starts?

Reply:
Just to think this through a bit more...

When you can `cancel`, that *same* thread will be used to run any follow actions added via `whenComplete`, `thenApply` and friends.  So if you call `WorldStateDonwloader.cancel` by the time that method returns we do actually know that `handleCancellation` has completed.  However, we don't know if `handleCancellation` runs before or after anything else that was chained to that future.

I think that will probably work out ok but we may want to have `createFuture` return the future returned from `whenComplete` instead of the original to fix it (just need to make sure it doesn't interfere with whether the returned future completes exceptionally or not).

Reply:
Another good catch.  I think I broke this when I removed the `status` var.  I reverted that change and now you shouldn't be able to execute `run()` before the cancellation has finished because the checks are now based on the `status` variable which gets updated inside of the synchronized `handleCancellation` method.

Reply:
I went ahead and merged because I have another branch built on top of this one.  But if you see any issues with my follow-up updates, I can fix them in another PR.